uc-crux-llvm/src/UCCrux/LLVM/Specs/Apply.hs

```diff
@@ -0,0 +1,184 @@
+-- uc-crux-llvm
+import           UCCrux.LLVM.Check (checkConstrainedShapes, SomeCheckedConstraint(..), CheckedConstraint(checkedPred))
+import           UCCrux.LLVM.Context.Module (ModuleContext)
+import           UCCrux.LLVM.ExprTracker (ExprTracker)
+import qualified UCCrux.LLVM.FullType.FuncSig as FS
+import           UCCrux.LLVM.FullType.Type (FullTypeRepr, MapToCrucibleType)
+import           UCCrux.LLVM.Module (FuncSymbol)
+import           UCCrux.LLVM.Postcondition.Apply (applyPostcond)
+import qualified UCCrux.LLVM.Postcondition.Type as Post
+import           UCCrux.LLVM.Specs.Type (Specs)
+import qualified UCCrux.LLVM.Specs.Type as Spec
+{- ORMOLU_ENABLE -}
+
+-- | Collapse all the preconditions of a 'Spec' to a single predicate
+matchPreconds ::
+  IsSymInterface sym =>
+  HasLLVMAnn sym =>
+  ArchOk arch =>
+  (?memOpts :: MemOptions) =>
+  ModuleContext m arch ->
+  sym ->
+  MemImpl sym ->
+  Ctx.Assignment (FullTypeRepr m) argTypes ->
+  Spec.SpecPreconds m argTypes ->
+  Ctx.Assignment (Crucible.RegEntry sym) (MapToCrucibleType arch argTypes) ->
+  IO (What4.Pred sym)
+matchPreconds modCtx sym mem argFTys preconds args =
+  do constraints <-
+       checkConstrainedShapes modCtx sym mem argFTys (Spec.specArgPreconds preconds) args
+     What4.andAllOf
+       sym
+       (Lens.folding (fmap (\(SomeCheckedConstraint c) -> checkedPred c)))
+       constraints
+
+-- | Apply a spec postcondition, modifying memory and creating a return value
+--
+-- If the postcondition of the spec is 'Nothing', apply 'Post.minimalPostcond'.
+applyPost ::
+  forall m arch sym bak fs va mft args argTypes p ext r cargs ret.
+  Crucible.IsSymBackend sym bak =>
+  IsSymInterface sym =>
+  HasLLVMAnn sym =>
+  ArchOk arch =>
+  (?lc :: TypeContext) =>
+  (?memOpts :: MemOptions) =>
+  (fs ~ 'FS.FuncSig va mft args) =>
+  bak ->
+  ModuleContext m arch ->
+  -- | Track origins of created values
+  IORef (ExprTracker m sym argTypes) ->
+  -- | Function name
+  FuncSymbol m ->
+  -- | Function type
+  FS.FuncSigRepr m fs ->
+  -- | LLVM memory variable
+  Crucible.GlobalVar Mem ->
+  Spec.Spec m fs ->
+  -- | Arguments to function
+  Ctx.Assignment (Crucible.RegEntry sym) (MapToCrucibleType arch args) ->
+  Crucible.OverrideSim p sym ext r cargs ret
+    (Crucible.RegValue sym (FS.ReturnTypeToCrucibleType arch mft))
+applyPost bak modCtx tracker funcSymb fsRep mvar spec args =
+  do FS.FuncSigRepr _ _ retTy <- return fsRep
+     Spec.Spec _ _ specPost _ <- return spec
+     let post = fromMaybe (Post.minimalPostcond fsRep) specPost
+     Crucible.modifyGlobal mvar $ \mem ->
+       liftIO (applyPostcond bak modCtx tracker funcSymb post retTy mem args)
+
+-- | Apply a collection of specs (a 'Specs') to the current program state.
+--
+-- Creates one symbolic branches per spec, as described on the Haddock for
+-- 'Specs' the semantics is that the first spec with a matching precondition has
+-- its postcondition applied to mutate memory and supply a return value.
```

Comment:
My worry is that the documentation for `symbolicBranches` is unclear and that we might be misinterpreting it. I'd like to ask Rob for a clarification. I *think* that the "first one that evaluates to true" is only referring to a predicate that is *concretely* true. If none of the predicates are concretely/syntactically `true`, you'll get a mux of all of the symbolic values.

Which isn't really a bad thing, but I think we need to clarify what `symbolicBranches` actually does and how this code makes use of it.

I'm building this list of worries based on the documentation of `symbolicBranch`, which muxes together both branches unless one is syntactically false/true.

Reply:
Speaking off the cuff here, but I'm pretty sure this is right. Semantically, `symbolicBranches` is basically an if/then/else chain, so the "first one that evaluates true" makes sense even in the symbolic case. Later branches will assume that earlier branches were not taken.  Operationally, yes, this will stop evaluating if one is concretely true, and will otherwise mux together all the branches that are not "obviously infeasible."  Even in the symbolic case, later branches assume earlier branches were not taken.

However, you should be able to emulate a truly nondeterministic choice by inventing a fresh variable and using it as part of the branch condition.  For example, number all of your branches from 0 to n, invent a symbolic integer variable `x`, and add `x = c` (where c ranges from 0 to n) as part of the branch condition for each branch. This should have the effect of generating a truly nondeterministic branch between the feasible paths.

Reply:
> Semantically, symbolicBranches is basically an if/then/else chain, so the "first one that evaluates true" makes sense even in the symbolic case. Later branches will assume that earlier branches were not taken.

Thanks for the clarification! I'll add something along these lines to the Haddock comment for `symbolicBranches`.

> However, you should be able to emulate a truly nondeterministic choice by inventing a fresh variable and using it as part of the branch condition. For example, number all of your branches from 0 to n, invent a symbolic integer variable x, and add x = c (where c ranges from 0 to n) as part of the branch condition for each branch. This should have the effect of generating a truly nondeterministic branch between the feasible paths.

Rad! Does a nondeterministic branching operation along these lines seem worth having in the `crucible` library? Otherwise, I can put it in `uc-crux-llvm`.

Reply:
Seems to me like that is useful enough to push into `crucible` proper.